PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,508 @@
+        preventPurchasePopupCallFromTriggeringCacheRefresh(appUserID: appUserID)
+
+        if let presentedOfferingIdentifier = maybePresentedOfferingIdentifier {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product_from_package,
+                                   productIdentifier,
+                                   presentedOfferingIdentifier))
+            lock.lock()
+            presentedOfferingIDsByProductID[productIdentifier] = presentedOfferingIdentifier
+            lock.unlock()
+
+        } else {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product, productIdentifier))
+        }
+
+        productsManager.cacheProduct(product)
+
+        lock.lock()
+        defer {
+            lock.unlock()
+        }
+
+        guard purchaseCompleteCallbacksByProductID[productIdentifier] == nil else {
+            completion(nil, nil, ErrorUtils.operationAlreadyInProgressError(), false)
+            return
+        }
+        purchaseCompleteCallbacksByProductID[productIdentifier] = completion
+        storeKitWrapper.add(payment)
+    }
+
+}
+
+extension PurchasesOrchestrator: StoreKitWrapperDelegate {
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                updatedTransaction transaction: SKPaymentTransaction) {
+        switch transaction.transactionState {
+        case .restored, // for observer mode
+             .purchased:
+            handlePurchasedTransaction(transaction)
+        case .purchasing:
+            break
+        case .failed:
+            handleFailedTransaction(transaction)
+        case .deferred:
+            handleDeferredTransaction(transaction)
+        @unknown default:
+            Logger.warn("unhandled transaction state!")
+        }
+    }
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                removedTransaction transaction: SKPaymentTransaction) {
+        // todo: remove
+        // unused for now
+    }
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                shouldAddStorePayment payment: SKPayment,
+                                for product: SKProduct) -> Bool {
+        productsManager.cacheProduct(product)
+
+        guard let delegate = maybeDelegate else { return false }
+        delegate.shouldPurchasePromoProduct(product) { completion in
+            self.purchaseCompleteCallbacksByProductID[product.productIdentifier] = completion
+            storeKitWrapper.add(payment)
+        }
+        return false
+    }
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
+                                didRevokeEntitlementsForProductIdentifiers productIdentifiers: [String]) {
+        Logger.debug(String(format: Strings.purchase.entitlements_revoked_syncing_purchases, productIdentifiers))
+        syncPurchases { _, _ in
+            Logger.debug(Strings.purchase.purchases_synced)
```

Comment:
FFTI: we should probably have some logging in `syncPurchases` itself, since we probably always want to log it

Reply:
That func is pretty complex and there would be multiple places where we'd need to log. I think this is better left alone for now.